### PR TITLE
Build Wheel Action

### PR DESCRIPTION
**Pull request by jbusche**
### Description of the change
This PR adds build and publish Github Actions that will allow extra testing of the build and wheel files with tox and then ultimately publish the library out on the production PyPI when we get the token figured out.

### Related issue number

Closes issue [671](https://github.ibm.com/ai-foundation/watson-fm-stack-tracker/issues/671) and external issue https://github.com/foundation-model-stack/fms-hf-tuning/issues/57

### How to verify the PR

From my [github repo](https://github.com/jbusche/fms-hf-tuning/tree/jb-wheel-issue57), I synced my branch with main and then am able to manually run the GHA successfully, giving it the input version, for example `1.0.dev3` in this run:  https://github.com/jbusche/fms-hf-tuning/actions/runs/8547067789

I tried then installing from a new python env:
```
pip install fms-hf-tuning==1.0.dev3
pip install fms-hf-tuning[dev]==1.0.dev3
pip list |grep fms
fms-hf-tuning            1.0.dev3
```

### Was the PR tested
Yes, by running the GHA in my fork.  (The GHA won't run in the upstream repo until it's synced with main branch)

- [ ] I have added >=1 unit test(s) for every new method I have added.
- [ ] I have ensured all unit tests pass

### CODE REVIEW

File: .github/workflows/build-library.yml
Comment (tedhtchang) [inline .github/workflows/build-library.yml:1]: This should be combined into 1 yaml

File: .github/workflows/publish-library.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up Python 3.9
+        uses: actions/setup-python@v4
+        with:
+          python-version: 3.9
+      - name: Build and check package
```
Comment (tedhtchang) [inline .github/workflows/publish-library.yaml:30]:
```suggestion
      - name: Build and test with tox
        run: tox -e py39
      - name: Build and check package
```

File: .github/workflows/build-library.yml
Comment (jbusche) [reply .github/workflows/build-library.yml:1]: OK, combined in one file, still have publish wheel commented out for now.  Thanks @tedhtchang

File: .github/workflows/publish-library.yaml (hunk 1 above)
Comment (jbusche) [reply .github/workflows/publish-library.yaml]: ok, combined into one file, thank you!

File: .github/workflows/publish-library.yaml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+name: Publish FMS-hf-tuning Library
+
+on:
+  release:
+    types: [published]
```
[the remark this answers is not in the document]
Comment (jbusche) [reply .github/workflows/publish-library.yaml]: Added to the top of build-and-publish.yaml, thank you!

File: .github/workflows/build-and-publish.yaml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        uses: actions/setup-python@v4
+        with:
+          python-version: 3.9
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          python -m pip install -r setup_requirements.txt
+      - name: Build and test with tox
+        run: tox -e py39
+      - name: Build and check wheel package
```
Comment (tedhtchang) [inline .github/workflows/build-and-publish.yaml:40]: Add a step before the  `Build and check wheel package` step to substitute version = "0.0.1"   with value="{{ github.event.inputs.release-version }}" in the pyproject.toml

Comment (jbusche) [reply .github/workflows/build-and-publish.yaml:49]: Done, thanks Ted

**Comment by jbusche** [comment]
OK this GHA works, building the wheel file and pushing up to PyPI.

Some questions remain:
1.  Is manually running the GHA and giving it the version number the right way to publish to PyPI or is it wanted to be automatic, for example when branch name = release or tag or something?  If so, how do we want to increment the version number... be sure to update the toml file prior to pushing the release branch, or maybe from an individual version file?
2. With the trusted upload to pypi, there is an "environment name" option. (see [doc](https://docs.pypi.org/trusted-publishers/adding-a-publisher/)) It's recommended to set for extra security, but I don't know what the environment name would be or how to set that.
3. In testing pushed 0.0.1 already, didn't realize it's not deletable.  I need to research if, while not deletable if it's updatable with newer wheel file(s) on the same version.

**Comment by tedhtchang** [comment]
@ z103cb Please advice.

**Comment by tedhtchang** [comment]
@z103cb Please advice.

**Review by z103cb — COMMENTED** [review]
I think that this PR is ok.  We should consider perhaps having this triggered from "Create release" page in GITHUB.  An example from RH library I have collaborated on can be found here: https://github.com/opendatahub-io/caikit-nlp-client/blob/main/.github/workflows/release.yml

File: .github/workflows/build-and-publish.yaml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        with:
+          python-version: 3.9
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          python -m pip install -r setup_requirements.txt
+      - name: Replace release-version with provided input value
+        run: |
+          export RELEASE_VERSION=${{ github.event.inputs.release-version }}
+          sed -i "s/^version = \"[0-9.]*\"$/version = \"${RELEASE_VERSION}\"/g" pyproject.toml
```
Comment (z103cb) [inline .github/workflows/build-and-publish.yaml:46]: Perhaps we should commit this change to the file or conversely, do it in the way, it's done the [here](https://github.com/opendatahub-io/caikit-nlp-client/blob/main/pyproject.toml).

**Comment by z103cb** [comment]
> OK this GHA works, building the wheel file and pushing up to PyPI.
> 
> Some questions remain:
> 
> 1. Is manually running the GHA and giving it the version number the right way to publish to PyPI or is it wanted to be automatic, for example when branch name = release or tag or something?  If so, how do we want to increment the version number... be sure to update the toml file prior to pushing the release branch, or maybe from an individual version file?

I would prefer it be done automatically, when a new "GitHub release" is created.  In the [example](https://github.com/opendatahub-io/caikit-nlp-client/blob/main/pyproject.toml) I have provided, believe that the version is derived from the tag on the release. The relevant bits are in:

```toml
[build-system]
build-backend = "setuptools.build_meta"
requires = [
    "setuptools>=60",
    "setuptools-scm>=8.0",
]


[project]
name = "caikit_nlp_client"
dynamic = ["version"]
description = "caikit-nlp client"
license = { text = "Apache License 2.0" }
readme = "README.md"
classifiers=[
    "License :: OSI Approved :: Apache Software License",
    "Development Status :: 4 - Beta",
    "Programming Language :: Python :: 3",
    "Programming Language :: Python :: 3.9",
    "Programming Language :: Python :: 3.10",
    "Programming Language :: Python :: 3.11",
]
```

> 2. With the trusted upload to pypi, there is an "environment name" option. (see [doc](https://docs.pypi.org/trusted-publishers/adding-a-publisher/)) It's recommended to set for extra security, but I don't know what the environment name would be or how to set that.
I would ask the owners of the repo

> 3. In testing pushed 0.0.1 already, didn't realize it's not deletable.  I need to research if, while not deletable if it's updatable with newer wheel file(s) on the same version.
I would not worry too much to about it.

File: .github/workflows/build-and-publish.yaml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+jobs:
+  build:
+    runs-on: ubuntu-20.04
+    environment:
+      name: pypi
+      url: https://pypi.org/p/fms-hf-tuning
+    permissions:
+      id-token: write  # IMPORTANT: this permission is mandatory for trusted publishing
+    steps:
+      - uses: actions/checkout@v3
```
Comment (zdtsw) [inline .github/workflows/build-and-publish.yaml:34]:
```suggestion
      - uses: actions/checkout@v4
```

File: .github/workflows/build-and-publish.yaml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    workflow_dispatch:
+        inputs:
+            release-version:
+              type: string
+              required: true
+              description: 'Version number (for example: 0.0.1)'
+
+jobs:
+  build:
+    runs-on: ubuntu-20.04
```
Comment (zdtsw) [inline .github/workflows/build-and-publish.yaml:27]:
```suggestion
    runs-on: ubuntu-latest
```
unless you have special requirement must run on 20.04

File: .github/workflows/build-and-publish.yaml (hunk 4 above)
Comment (zdtsw) [reply .github/workflows/build-and-publish.yaml:46]: i could not find `version =`  is matched in [pyproject.toml](https://github.com/opendatahub-io/caikit-nlp-client/blob/main/pyproject.toml)

**Review by zdtsw — COMMENTED** [review]
some nitty gritty for the GH action part

File: .github/workflows/build-and-publish.yaml (hunk 4 above)
Comment (tedhtchang) [reply .github/workflows/build-and-publish.yaml:46]: you may find `version = ` in this [pyproject.toml](https://github.com/foundation-model-stack/fms-hf-tuning/blob/445c17b7c0bc6bd87794334d2eab71b495382f2b/pyproject.toml#L8)

**Comment by tedhtchang** [comment]
>I would prefer it be done automatically, when a new "GitHub release" is created. In the [example](https://github.com/opendatahub-io/caikit-nlp-client/blob/main/pyproject.toml) I have provided, believe that the version is derived from the tag on the release.

@z103cb This a good suggestion. Btw, the version comes from a [file](https://github.com/opendatahub-io/caikit-nlp-client/blob/df27724b70a1cd048fc833844cda8d5e215d1f0c/pyproject.toml#L45).

File: .github/workflows/build-and-publish.yaml [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+name: Build and Publish FMS-hf-tuning Library
+
+on:
```
Comment (tedhtchang) [inline .github/workflows/build-and-publish.yaml:17]: In addition to the change in  pyproject.toml also add
```suggestion
on:
    release:
      types: [published]
```

File: .github/workflows/build-and-publish.yaml [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    workflow_dispatch:
+        inputs:
+            release-version:
+              type: string
+              required: true
+              description: 'Version number (for example: 0.0.1)'
```
Comment (tedhtchang) [inline .github/workflows/build-and-publish.yaml:23]: Remove

File: .github/workflows/build-and-publish.yaml [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+      - uses: actions/checkout@v3
+      - name: Set up Python 3.9
+        uses: actions/setup-python@v4
+        with:
+          python-version: 3.9
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          python -m pip install -r setup_requirements.txt
+      - name: Replace release-version with provided input value
```
Comment (tedhtchang) [inline .github/workflows/build-and-publish.yaml:43]: remove

File: .github/workflows/build-and-publish.yaml (hunk 4 above)
Comment (tedhtchang) [inline .github/workflows/build-and-publish.yaml:46]: remove

File: .github/workflows/build-and-publish.yaml (hunk 8 above)
Comment (tedhtchang) [inline .github/workflows/build-and-publish.yaml:23]: maybe removed

File: .github/workflows/build-and-publish.yaml [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        run: |
+          export RELEASE_VERSION=${{ github.event.inputs.release-version }}
+          sed -i "s/^version = \"[0-9.]*\"$/version = \"${RELEASE_VERSION}\"/g" pyproject.toml
+      - name: Build and test with tox
+        run: tox -e py39
+      - name: Build and check wheel package
+        run:
+          tox -e build,twinecheck
+      - name: Publish package distributions to PyPI
+        uses: pypa/gh-action-pypi-publish@release/v1
```
Comment (gkumbhat) [inline .github/workflows/build-and-publish.yaml:53]: We don't need any password or token for this?

Comment (gkumbhat) [reply .github/workflows/build-and-publish.yaml:53]: We can also restrict this to releases.

File: .github/workflows/build-and-publish.yaml (hunk 6 above)
Comment (jbusche) [reply .github/workflows/build-and-publish.yaml:27]: I originally had it as ubuntu-latest but I was seeing warnings that the older nodejs was deprecated and we should move to 20... I can revert back to ubuntu-latest and let it ride for now, thanks!

File: .github/workflows/build-and-publish.yaml [hunk 11]
```diff
@@ -15,37 +15,39 @@
...
+      types: [published]
 
 jobs:
   build:
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version:
+          # NOTE: TGIS does not support 3.8 or 3.11
```
Comment (gkumbhat) [inline .github/workflows/build-and-publish.yaml:27]: I believe TGIS does support `3.11` ..  `3.8` is deprecated, so we don't need to support that.

Comment (jbusche) [reply .github/workflows/build-and-publish.yaml:27]: Thanks @gkumbhat... I'll add 3.11.  I've made a lot of changes and it's stuck failing test ATM, trying to figure out what's wrong...

Comment (jbusche) [reply .github/workflows/build-and-publish.yaml]: OK - added, thanks!

File: .github/workflows/build-and-publish.yaml (hunk 10 above)
Comment (jbusche) [reply .github/workflows/build-and-publish.yaml:56]: No passwod.token needed for now, as we're using trusted upload to pypi, (see [doc](https://docs.pypi.org/trusted-publishers/adding-a-publisher/)) but we might want to change this to a user/token.

I think I've set it to release now, so have removed the sed section

File: .github/workflows/build-and-publish.yaml [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+      name: pypi
+      url: https://pypi.org/p/fms-hf-tuning
+    permissions:
+      id-token: write  # IMPORTANT: this permission is mandatory for trusted publishing
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up Python 3.9
+        uses: actions/setup-python@v4
+        with:
+          python-version: 3.9
```
[the remark this answers is not in the document]
Comment (jbusche) [reply .github/workflows/build-and-publish.yaml]: OK, I have it as a matrix, hoping that covers 3.9-3.11

File: .github/workflows/build-and-publish.yaml (hunk 8 above)
Comment (jbusche) [reply .github/workflows/build-and-publish.yaml]: Removed, thanks Ted

File: .github/workflows/build-and-publish.yaml (hunk 4 above)
Comment (jbusche) [reply .github/workflows/build-and-publish.yaml]: Removed, thanks

File: .github/workflows/build-and-publish.yaml (hunk 7 above)
Comment (jbusche) [reply .github/workflows/build-and-publish.yaml:17]: Added, thanks

File: .github/workflows/build-and-publish.yaml (hunk 6 above)
Comment (jbusche) [reply .github/workflows/build-and-publish.yaml]: switched back to `runs-on: ubuntu-latest`  thanks

File: .github/workflows/build-and-publish.yaml (hunk 5 above)
Comment (jbusche) [reply .github/workflows/build-and-publish.yaml]: changed to v4, thanks!

File: .github/workflows/build-and-publish.yaml (hunk 4 above)
Comment (jbusche) [reply .github/workflows/build-and-publish.yaml]: OK - working on this... thanks for the advice, reviews!

**Comment by jbusche** [comment]
@gkumbhat and @tedhtchang, thanks for the suggestions, I got it to work on `release` where it builds the wheel and publishes up to PiPY.  I just tried with 0.0.1rc.5 from my repo (had to use my main bracn so it would work):
https://github.com/jbusche/fms-hf-tuning/actions/runs/8620146700/job/23626417173
```
pip install fms-hf-tuning==0.0.1rc.5
pip install fms-hf-tuning[dev]==0.0.1rc.5
```
I think the only thing perhaps left is that people didn't want it set to just python 3.9, so I added a matrix that added all three in a manor like I saw in [caitkit-nlp](https://github.com/caikit/caikit-nlp/blob/main/.github/workflows/build-library.yml#L28-L33)  However having 3.9, 3.10 and 3.11 ended up launching three identical workflows which isn't what I wanted, so I've made it just 3.11.  Is that sufficient?

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -35,6 +35,13 @@ We require new unit tests to be contributed with any new functionality added.
 
 Before sending pull requests, make sure your changes pass formatting, linting and unit tests. These checks will run with the pull request builds. Alternatively, you can run the checks manually on your local machine [as specified below](#development).
 
+#### Dependencies
+If additional new Python module dependencies are required, think about where to put them:
+
+- If they're required for fms-hf-tuning, then append them to the end of the requirements.txt file.
```
Comment (tedhtchang) [inline CONTRIBUTING.md:41]: We want to deprecate the requirements.txt file. Let's use pyproject.toml to manage dependencies.

File: tox.ini [hunk 14]
```diff
@@ -6,6 +6,7 @@ description = run unit tests
 deps =
     pytest>=7
     -r requirements.txt
+    wheel>=0.38.4
```
Comment (tedhtchang) [inline tox.ini:9]: Use the dev instead. This will also let you pass the lint error
```suggestion
    .[dev]
```

Comment (jbusche) [reply tox.ini:9]: Thanks @tedhtchang, we'll see if the lint error passes now, but I think @alex-jw-brooks fixed this when he included `.[dev]` in the tox.ini file for his [PR 117](https://github.com/foundation-model-stack/fms-hf-tuning/pull/117)

File: CONTRIBUTING.md (hunk 13 above)
Comment (jbusche) [reply CONTRIBUTING.md:41]: Thank you Ted, I will adjust ASAP once we decide to switch entirely to the toml file.  Really the only use-case I see for keeping the requirements.txt file is if a developer wanted all the pre-reqs without installing the fms-hf-tuning module yet.  If that's not real use-case, then the toml file is a lot cleaner way to manage the dependencies.

**Comment by jbusche** [comment]
PR 105 is looking good in testing from my main branch, testing creating a new release.  It builds the wheel now without source (thanks for the tip @TED CHANG on how to only build the wheel file) and publishes up to PiPY.  Can see that only the wheel file is published:  https://pypi.org/project/fms-hf-tuning/0.0.1rc6/#files

Tried on a fresh python environment, and it's looking good:
```
pip install --upgrade pip
pip install fms-hf-tuning==0.0.1rc.6
pip install fms-hf-tuning[dev]==0.0.1rc.6
```
and the install looks good:
```
pip list |grep fms
fms-hf-tuning            0.0.1rc6
```
and I was successfully able to run the twitter_complaints test
```
100%|██████████████████████████████████████████████████████████████████████████████████████████| 3/3 [01:33<00:00, 31.08s/it]
```

File: tox.ini (hunk 14 above)
Comment (tedhtchang) [reply tox.ini:9]: Do you need 'wheel' to run pytest ?

**Review by tedhtchang — APPROVED** [review]
/LGTM Thanks.

File: setup.py
Comment (gkumbhat) [inline setup.py:1]: can we remove the need for this file? As per my suggestion here: https://github.com/foundation-model-stack/fms-hf-tuning/pull/68#pullrequestreview-1962018424

Comment (jbusche) [reply setup.py:1]: OK, created a new [issue 756](https://github.ibm.com/ai-foundation/watson-fm-stack-tracker/issues/756) where I'll try to eliminate the setup.py file as suggested.   Thanks!

File: pyproject.toml [hunk 15]
```diff
@@ -29,6 +33,13 @@ aim = ["aim==3.18.1"]
 [tool.setuptools.dynamic]
 dependencies = {file = ["requirements.txt"]}
 
+[tool.setuptools.packages.find]
+exclude = ["tests", "tests.*"]
+namespaces = false
+
+[tool.setuptools_scm]
+version_file = "tuning/_version.py"
```
[the remark this answers is not in the document]
Comment (jbusche) [reply pyproject.toml:41]: Right, it's generated... it's the same with caikit-nlp and caikitt-nlp-client, they have:
https://github.com/caikit/caikit/blob/main/pyproject.toml#L132
and
https://github.com/opendatahub-io/caikit-nlp-client/blob/df27724b70a1cd048fc833844cda8d5e215d1f0c/pyproject.toml#L45

But no actual `_version.py` file in their paths

**Review by gkumbhat — APPROVED** [review]

**Pull request merged**